### DISCUSSION TITLE
Crashes with VS 2015 project (.NET Framework 4.6.1)

### DISCUSSION THREAD

**Discussion by Th-69** [dotnet/Silk.NET#311] [General] [answered]
# Summary
I use the newest Silk.NET NuGet package (version 1.6.0) and created a VS 2015 project with it (OpenGL).
If I run the debugger inside VS 2015 then the window appears, but after 1 - 2 seconds the process crashes with
"The program '[12436] ExampleApp.exe' has exited with code -1073740791 (0xc0000409)." -> corrupted stacktrace

I can debug the project until "window.Initialize()" or "window.Run()" is called, after that the debugger stops with "Application is in break mode" with
"Managed Debugging Assistant 'FatalExecutionEngineError' has detected a problem in 'ExampleApp\bin\Debug\ExampleApp.exe'.

Additional information: Die Laufzeit hat einen schwerwiegenden Fehler entdeckt. Fehleradresse: "0x6f510c11" in Thread "0x47d8". Fehlercode: 0xc0000005." ("The runtime has detected a critical error").

But If I run the program outside of VS (from Windows Explorer) the process is stable (and runs normally).

And the second crash appears if I call (inside Load event):
"IInputContext input = window.CreateInput();" (without any more code, e.g. binding key or mouse events - but even with it it crashes).

I can debug (step over) that line, but after showing the window the process crashes with
"The program '[20044] ExampleApp.exe' has exited with code -1073740771 (0xc000041d)." -> STATUS_FATAL_USER_CALLBACK_EXCEPTION

And outside of VS the same happens, the process crashes (but without any message).

So I can't use any input...

# Steps to reproduce
- Platform: Desktop
- Framework Version: .NET 4.6.1
- API: OpenGL
- API Version: OpenGL 3.3
s. above

# Comments

**Comment by Perksey** [comment]
The first error I'm absolutely clueless as to why that's happening, but the second error _seems_ like a really rare bug encountered with the 1.X input backend where some memory we need is freed prematurely by the runtime, try one of the 2.0 previews to see if you can reproduce this.

**Comment by Th-69** [comment]
Thank you for your fast response.

After updating the NuGet references to "2.0.0-preview2" (and changing some method parameters) "window = Window.Create(options);" now throws an PlatformNotSupportedException:
"An unhandled exception of type 'System.PlatformNotSupportedException' occurred in Silk.NET.Windowing.Common.dll"

Btw: Do I understand it correctly (https://github.com/Ultz/Silk.NET/pull/310), that you also will (or have) implement(ed) a Direct3D binding for Silk.NET 2.0?

**Comment by Perksey** [comment]
Interesting. What packages are you referencing? In Silk.NET 2.0, we've moved to a soft dependency model instead of our previous "stiching assembly" model. Also do you know the message and stack trace of the PlatformNotSupportedException?

---

Yeah, per our 2.0 goals outlined in #48, #310 introduces DirectX bindings. Currently in that PR, there are some complete (but untested currently) bindings to Direct3D 9, 11, and 12.

**Comment by Th-69** [comment]
This is the content of the "packages.config":
```
<?xml version="1.0" encoding="utf-8"?>
<packages>
  <package id="Microsoft.Bcl.HashCode" version="1.1.0" targetFramework="net461" />
  <package id="Microsoft.CSharp" version="4.7.0" targetFramework="net461" />
  <package id="Microsoft.DotNet.PlatformAbstractions" version="3.1.6" targetFramework="net461" />
  <package id="Microsoft.Extensions.DependencyModel" version="3.1.1" targetFramework="net461" />
  <package id="Newtonsoft.Json" version="9.0.1" targetFramework="net461" />
  <package id="Silk.NET" version="2.0.0-preview2" targetFramework="net461" />
  <package id="Silk.NET.Core" version="2.0.0-preview2" targetFramework="net461" />
  <package id="Silk.NET.GLFW" version="2.0.0-preview2" targetFramework="net461" />
  <package id="Silk.NET.Input" version="2.0.0-preview2" targetFramework="net461" />
  <package id="Silk.NET.Input.Common" version="2.0.0-preview2" targetFramework="net461" />
  <package id="Silk.NET.Input.Glfw" version="2.0.0-preview2" targetFramework="net461" />
  <package id="Silk.NET.Input.Sdl" version="2.0.0-preview2" targetFramework="net461" />
  <package id="Silk.NET.OpenAL" version="2.0.0-preview2" targetFramework="net461" />
  <package id="Silk.NET.OpenGL" version="2.0.0-preview2" targetFramework="net461" />
  <package id="Silk.NET.SDL" version="2.0.0-preview2" targetFramework="net461" />
  <package id="Silk.NET.Vulkan" version="2.0.0-preview2" targetFramework="net461" />
  <package id="Silk.NET.Vulkan.Extensions.KHR" version="2.0.0-preview2" targetFramework="net461" />
  <package id="Silk.NET.Windowing" version="2.0.0-preview2" targetFramework="net461" />
  <package id="Silk.NET.Windowing.Common" version="2.0.0-preview2" targetFramework="net461" />
  <package id="Silk.NET.Windowing.Desktop" version="1.6.0" targetFramework="net461" />
  <package id="Silk.NET.Windowing.Glfw" version="2.0.0-preview2" targetFramework="net461" />
  <package id="Silk.NET.Windowing.Sdl" version="2.0.0-preview2" targetFramework="net461" />
  <package id="System.Buffers" version="4.5.1" targetFramework="net461" />
  <package id="System.Memory" version="4.5.4" targetFramework="net461" />
  <package id="System.Numerics.Vectors" version="4.5.0" targetFramework="net461" />
  <package id="System.Reflection.Emit" version="4.7.0" targetFramework="net461" />
  <package id="System.Reflection.Emit.ILGeneration" version="4.7.0" targetFramework="net461" />
  <package id="System.Runtime.CompilerServices.Unsafe" version="4.7.1" targetFramework="net461" />
  <package id="System.Runtime.InteropServices.RuntimeInformation" version="4.0.0" targetFramework="net461" />
  <package id="Ultz.Bcl.Half" version="1.0.0" targetFramework="net461" />
  <package id="Ultz.Bcl.Half.Fallback" version="1.0.0" targetFramework="net461" />
  <package id="Ultz.Native.GLFW" version="3.3.2" targetFramework="net461" />
  <package id="Ultz.Native.SDL" version="2.0.12-pre2" targetFramework="net461" />
  <package id="Ultz.SuperInvoke" version="1.1.0" targetFramework="net461" />
  <package id="Ultz.SuperInvoke.Loader" version="1.1.0" targetFramework="net461" />
</packages>
```

The exception message is just "Couldn't find a suitable window platform. https://docs.ultz.co.uk/Silk.NET/windowing/troubleshooting.html" (but this site is unavailable for me ;-( ).
And there's no further stack trace, only my own code:
- Silk.NET.Windowing.Window.Create(WindowOptions options)
- ExampleApp.Program.Main()

I've also tested it with VS 2017 and I get the same platform exception (after a full clean and rebuild).

What is exactly meant by "platform not supported"?
Isn't v2.0 running anymore with VS 2015 or VS 2017 (and .NET 4.6.1)?

I've developed an UI library and the Win GDI+ (WinForms) renderer is working and now I'm trying to implement other renderers (OpenGL, Vulkan, DirectX) for it. And if DirectX is supported in future with Silk.NET 2.0 I don't have to use another lib (and that would be great ;-).

I'm waiting for .NET 5 and the next Visual Studio version (like many other people), but for now I use VS 2015 (I don't want to install VS2019 and in a few month another new VS ;-).

PS: I'm getting the warning "The referenced component 'System.Runtime.InteropServices.RuntimeInformation' could not be found." (but also with Silk.NET 1.6, so I've ignored this warning so far).

**Comment by Perksey** [comment]
Could be related to #192

**Comment by Perksey** [comment]
Specify `<RuntimeIdentifier>` element in your csproj in order to have it copy the file into the bin dir.
<sub>_Original from @Tyler-IN in #192_</sub>

**Comment by Th-69** [comment]
The "glfw3.dll" (32-bit) is already in the bin/debug (and release) folder (tried it also with the 64 bit version). With 1.6 I got another exception for that: "GLFW: Copy a GLFW 3.3 binary into the same directory as ExampleApp".

And isn't `<RuntimeIdentifier>` only for .NET Core projects?

Edit: But even with `<RuntimeIdentifier>net461</RuntimeIdentifier>` in the csproj file I get the same platform exception.

**Comment by Perksey** [comment]
That's not the correct usage of RuntimeIdentifier, you want something like win-x64 in there.

If that doesn't work you'll have to see if you can get more insight. Specifically, if any first chance exceptions are being thrown and caught silently (they'll contain valuable info) and if Silk.NET.Windowing.Glfw.dll is being loaded or not.

**Comment by Th-69** [comment]
I've now tried with "win", "win-x32", "win-x64" and even "win10", "win10-x32" and "win10-x64" (according to [Windows RIDs](https://docs.microsoft.com/en-us/dotnet/core/rid-catalog#windows-rids)), but all give the platform exception.

But just for clarification: I've put the `<RuntimeIdentifier>XXX</RuntimeIdentifier>` in the first `<PropertyGroup>` of the csproj-file:
```
<Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props" Condition="Exists('$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props')" />
  <PropertyGroup>
    <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
    <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
    <ProjectGuid>{B18912CC-AC84-4A8C-901F-8730B199B269}</ProjectGuid>
    <OutputType>Library</OutputType>
    <AppDesignerFolder>Properties</AppDesignerFolder>
    <RootNamespace>Win3D</RootNamespace>
    <AssemblyName>Win3D</AssemblyName>
    <TargetFrameworkVersion>v4.6.1</TargetFrameworkVersion>
    <FileAlignment>512</FileAlignment>
    <ImplicitlyExpandDesignTimeFacades>false</ImplicitlyExpandDesignTimeFacades>
    <TargetFrameworkProfile />
    <RuntimeIdentifier>win</RuntimeIdentifier>
  </PropertyGroup>
...
```
and then rebuilded and started the "Debug" configuration.

And in the "Module" list, only (from Silk.NET libraries)
- Silk.NET.Core.dll
- Silk.NET.Windowing.Common.dll

are loaded.
But I don't see any first chance exceptions in the debug output.

And in comparison with my running Win-GDI+ project, there are a lot more (> 20) entries of:
- Exception thrown: 'System.IO.FileNotFoundException' in mscorlib.dll (but it doesn't show the missing filename ;-( )

**Comment by Tyler-IN** [comment]
For older csproj, you may need \<Platform> / \<PlatformTarget> or something similar.

You should probably convert the project to use the ".NET Core" csproj format (which is not specific to .NET Core, it's just not backwards compatible - not sure if VS 2015 can support it) and \<TargetFramework>net461

If you want to upgrade your csproj, if you post your full csproj I can post what it should be converted to.
I should probably make a tool to do it at some point...

You can check the [UltralightSharp](https://github.com/ImpromptuNinjas/UltralightSharp) repo for an example that uses Silk.NET and see how the standard build process deploys the library in the bin dir. While the Silk.NET demo project is .NET Core, the project library itself is set up and tests (deploys properly) in CI on Linux/Mac/Win for .NET core as well as for .NET Fw under the GH windows runner just by normally building and running. Can check the Actions tab to see the results.

**Comment by Th-69** [comment]
Also with `<PlatformTarget>AnyCPU</PlatformTarget>` it is not running (with `x86` or `x64` I get a warning, cause my main project uses `AnyCPU`).

No, VS 2015 can't support directly .NET Core projects (it can use NuGet Standard 2.0 projects, but one cannot create or use .NET Core projects).
I can understand that Silk.NET 2.0 only uses the new project format and it would be sufficient for me to use v1.6.
I just want a running sample with OpenGL (and Vulkan) for my project.
Is there a workaround possible for that crash?

I've added my current csproj file (it is a class library):
[THX.Win3D.csproj.zip](https://github.com/Ultz/Silk.NET/files/5223153/THX.Win3D.csproj.zip)

Btw: This seems to be a really confusing thing: [Design/Doc issue: PlatformTarget vs Platforms vs RuntimeIdentifier handled inconsistently](https://github.com/dotnet/sdk/issues/1553)

**Comment by Th-69** [comment]
Now I reverted back to v1.6 and tried to use Vulkan:
```
var options = WindowOptions.Default;
options.API = GraphicsAPI.DefaultVulkan;
// ...
window = Window.Create(options);
```
But at `_gl = GL.GetApi(_window);` I get the exception:
"An unhandled exception of type 'System.InvalidOperationException' occurred in Silk.NET.OpenGL.dll
Additional information: The given IGLContextSource is not configured with a context."

`window.GLContext` is `null`.
How can I active Vulkan?

**Comment by Perksey** [comment]
You're trying to use OpenGL with a window configured for Vulkan.

**Comment by Th-69** [comment]
OK, I need something like `var vk = Vk.GetApi(...);`. Are there any examples for using Vulkan (as for OpenGL)?

And is there a common interface for both or do I have to develop different code for both APIs?

**Comment by Perksey** [comment, marked as answer]
We haven't published an official example, however there is a Vulkan test application in the src/Lab/VulkanTriangle directory 

You'll need to develop for both separately, Silk.NET is nothing more than a bindings library do you'll need to create the common API layer yourself
